### PR TITLE
introduce the cfregion lib to read and write file series in less space

### PR DESCRIPTION
**Pull request by kthielen**
This defines a basic compression option for stored series that accumulates a type-directed statistical model of recorded data to drive a basic arithmetic (entropy) encoder/decoder.

### CODE REVIEW

File: include/hobbes/cfregion.H [hunk 1]
```diff
@@ -0,0 +1,1181 @@
...
+  typedef uint32_t freq;
+
+  static const uint32_t cbits    = 16;
+  static const code     cmax     = (code(1) << cbits) - code(1);
+  static const code     cfourth  = code(1) << (cbits-2);
+  static const code     chalf    = 2*cfourth;
+  static const code     c3fourth = 3*cfourth;
+
+  static const uint32_t fbits = 14;
+  static const freq     fmax  = (freq(1) << fbits) - freq(1);
```
Comment (brianegge) [inline include/hobbes/cfregion.H]: Can these be made ```constexpr```?

Comment (kthielen) [reply include/hobbes/cfregion.H]: Yes good point, I will try that.

File: include/hobbes/cfregion.H [hunk 2]
```diff
@@ -0,0 +1,1181 @@
...
+      }
+      ++pm->freqs[s];
+      ++pm->c;
+    }
+  };
+
+/*******************************************************
+ *
+ * compress<T> : the main interface for type translation into slog compression data
+ *               this is specialized per-type to determine the model for accumulating stats, encoding entropy
```
Comment (brianegge) [inline include/hobbes/cfregion.H]: Neat

File: include/hobbes/cfregion.H [hunk 3]
```diff
@@ -0,0 +1,1248 @@
...
+  static constexpr code     cfourth  = code(1) << (cbits-2);
+  static constexpr code     chalf    = 2*cfourth;
+  static constexpr code     c3fourth = 3*cfourth;
+
+  static constexpr uint32_t fbits = 14;
+  static constexpr freq     fmax  = (freq(1) << fbits) - freq(1);
+
+  // bit patterns to shift in up to 16 ones
+  static code shift1s(code c) {
+    switch (c) {
```
Comment (brianegge) [inline include/hobbes/cfregion.H]: would ```(1 << c) -1``` be faster?

**Comment by kthielen** [comment]
Some rough details -- with these last couple of changes to improve performance, I ran a test program that recorded 10 million values of type '{x:int, y:byte, c:|Red,Green,Blue,Magenta|, |jimmy:int, bob:[char]|, nats:[long]}' both in "raw" form (the way we usually record data) and also in this "compressed" form.

The raw recording process took 15.4 seconds and produced a 1.4GB file.  Running gzip on this file took an additional 20.5 seconds and reduced the file to 109MB.  So a total of ~36 seconds and reduction to 7% of raw file size.

With this compression method, I recorded the same data in 26.3 seconds and produced a 158MB file.  So ~37% faster this way, but ~45% more bloat (than the final gzipped output).

I think that it's probably worth using this option intraday for very large files that are otherwise unwieldy when written/queried uncompressed (which was the reason for this PR).  But probably for most files, we are better off writing uncompressed.

**Comment by brianegge** [comment]
Would you humor me and measure the time required to bzip2 your 1.4GB sample?

**Comment by kthielen** [comment]
Strictly speaking, the escape probabilities here are overly conservative (implemented as <byte-value> / 256).  Since we will know that some of those 256 symbols are already accounted for in the non-escape set, we can reduce the escape symbol range proportionally.

In my tests, implementing escape probability this way only shrunk the compressed file by 0.7% (157MB instead of 158MB) but added 3% to run time (27.1s versus 26.3s) so probably isn't worth it.

**Comment by kthielen** [comment]
Hi Brian,

Sure, I ran bzip2 on the raw 1.4GB file and it took 1m35s and compressed down to 58MB.  So that's 322% slower than this method but 36% of the file size.

Also I ran xz on the raw 1.4GB file.  It took 7m8s and compressed down to 26MB.  So that's 1527% slower than this method but ~16% of the file size.

I'm not sure if there's a good way to make those methods incremental or support the other couple of requirements we have in this case (kill and resume the writing process, seek to batches quickly).

There's no doubt that this method is not optimal for final file size (even when compared to gzip, let alone bzip or xz).  It's pretty fast though, and gets our largest intraday file sizes down to a tolerable level -- we can still batch compress at the end of the day for optimal archive file sizes.

**Comment by brianegge** [comment]
Yes, I wasn't suggestion a different algo for on the fly, but we may want to consider using xz for offline compression.

**Comment by kthielen** [comment]
Agreed.

**Pull request merged**